core/src/main/java/org/apache/calcite/rel/core/Filter.java

```diff
@@ -85,6 +86,8 @@ protected Filter(
       RexNode condition) {
     super(cluster, traits, child);
     this.condition = requireNonNull(condition, "condition");
+    assert SqlTypeName.BOOLEAN == condition.getType().getSqlTypeName()
```

Comment:
I am not sure this is always true; for example, before validation where casts are introduced this may not hold.
Also, Calcite does not always represent SQL implicit casts explicitly, so the cast to Boolean which is implied here may not be in the program.

Reply:
+1

Reply:
The SQL level (SqlNode) and the algebraic level (Rel/RexNode) are handled differently. The validator as well as the implicit casts logic apply on the SQL level and not at the algebraic level. The algebra is explicit so the assertion for Boolean type makes sense to me.

Other than that it may be better to put the check inside the `isValid` method which is the canonical way of checking validity and allows also overriding in case some specialized implementation wants to perform the validity check differently.